### ISSUE TITLE
New Annotation to Apollo Specification

### ISSUE THREAD

**Issue by Nikhilkoneru** [apollographql/federation#428]
Hi Team,

From the query plan which I added below we can see that Apollo Gateway is waiting for SA(Service A) to respond even when we are provide the primary key(appNo) for the type inside the query as a param. We want to call  SA and SB parallelly when a client uses the below query, so I'm requesting to introduce a new annotation for Queries which can tell gateway that client params are primary keys of a query return type, I think this helps gateway to make calls parallely to SA and SB.

**SA - Schema**
```
type App @key(fields: "appNo" ) {
    appNo: Int!
    name: String!
}
type Query {
    getApp (appNo: String!): App
}
```
**SB - Schema**
```
type App @key(fields: "appNo" ) @extends {
    appNo: Int! @external
    description: String!
}
```
**Client Request to Apollo Gateway**
```
query {
  getApp(appNo: 1) {
   appNo
   name
   description
  }
}
```
**Query Plan of Gateway for the above request**
```
QueryPlan {
  Sequence {
    Fetch(service: "SA") {
      {
        getApp(appId: 1) {
          name
          __typename
          appId
        }
      }
    },
    Flatten(path: "getApp") {
      Fetch(service: "SB") {
        {
          ... on App {
            __typename
            appId
          }
        } =>
        {
          ... on App {
            description
          }
        }
      },
    },
  },
}
```

**Comment by Nikhilkoneru** [comment]
If accepted, I would love to contribute to this repository by adding a new annotation to @apollo/federation.

**Comment by Nikhilkoneru** [comment]
Trevor Scheer

**Comment by martijnwalraven** [comment]
This is definitely something we've thought about before. Not just in the context of Federation, but also to inform client-side caches (like Apollo Client).

One proposal we floated around was to annotate fields with a `@finder` directive that indicates a field obeys the expected semantics. By default, we'd assume a one-to-one correspondence between field arguments and key fields for example. (That also means we'd need the validation rules in place to check that.) If further customization is needed, perhaps we could expose a second directive (maybe something like `finder__map(to:)`) to map field arguments to a differently named key field.

Building this out isn't simple unfortunately, because it would require changes to composition and query planning. We're in the process of redesigning both, so even though we won't be able to get this done in the short term, I think this is worth picking up as part of that.

**Comment by ashu-walmart** [comment]
Makes sense. Is there any contribution that can be made to facilitate addition of the directives - finder/finder_mapTo.

**Comment by martijnwalraven** [comment]
> Makes sense. Is there any contribution that can be made to facilitate addition of the directives - finder/finder_mapTo.

Not really right now. Taking advantage of those depends on planned changes to composition and query planning, plus we are working on a more principled way of adding directives on subgraphs.

**Comment by codedawi** [comment]
+1 we would love this feature!
